File: 2839. check-if-strings-can-be-made-equal-with-operations-i.py
from collections import defaultdict

class Solution:
    def canBeEqual(self, s1: str, s2: str) -> bool:
        odd, even = defaultdict(int), defaultdict(int)
        for i in range(len(s1)):
            if i % 2 == 0:
                even[s1[i]] += 1
                even[s2[i]] -= 1
            else:
                odd[s1[i]] += 1
                odd[s2[i]] -= 1

        return set(odd.values()) == {0} and set(even.values()) == {0}

if __name__ == "__main__":
    sol = Solution()
    
    # Example 1
    s1_1 = "abcd"
    s2_1 = "cdab"
    output1 = sol.canBeEqual(s1_1, s2_1)
    print(f"Example 1: s1='{s1_1}', s2='{s2_1}'")
    print(f"Output: {output1}")
    print(f"Expected: True")
    print()
    
    # Example 2
    s1_2 = "abcd"
    s2_2 = "dacb"
    output2 = sol.canBeEqual(s1_2, s2_2)
    print(f"Example 2: s1='{s1_2}', s2='{s2_2}'")
    print(f"Output: {output2}")
    print(f"Expected: False")

File: 2839. check-if-strings-can-be-made-equal-with-operations-i.java
class Solution {
    public boolean canBeEqual(String s1, String s2) {
        char[] ch1 = s1.toCharArray();
        char[] ch2 = s2.toCharArray();
        int i = 0;
        while(i < ch1.length){
            if(ch1[i] == ch2[i]){
                i++;
            }
            else{
                if(i+2 < ch1.length){
                    char temp = ch1[i];
                    ch1[i] = ch1[i+2];
                    ch1[i+2] = temp;
                }
                i++;
            }
        }
        for(int k=0;k<ch1.length;k++){
            if(ch1[k] != ch2[k]) return false;
        }
        return true;
    }
    
    public static void main(String[] args) {
        Solution sol = new Solution();
        
        // Example 1
        String s1_1 = "abcd";
        String s2_1 = "cdab";
        boolean output1 = sol.canBeEqual(s1_1, s2_1);
        System.out.println("Example 1: s1='abcd', s2='cdab'");
        System.out.println("Output: " + output1);
        System.out.println("Expected: true");
        System.out.println();
        
        // Example 2
        String s1_2 = "abcd";
        String s2_2 = "dacb";
        boolean output2 = sol.canBeEqual(s1_2, s2_2);
        System.out.println("Example 2: s1='abcd', s2='dacb'");
        System.out.println("Output: " + output2);
        System.out.println("Expected: false");
    }
}